### PR TITLE
GH-113373: Speed up pathlib parsing by avoiding some attribute lookup misses

### PR DESCRIPTION
**Pull request by barneygale**
Improvements:

- Compute string representation in `_load_parts()`, which results in fewer attribute lookup misses.
- Store and re-use result of `os.path.join()`
- Check for the common case (`len(paths) == 1`) first, and speed up the subsequent `elif` by avoiding a length check

---

* Issue: gh-113373

### CODE REVIEW

**Comment by barneygale** [comment]
Some timings:

```bash
$ ./python -m timeit -n 500000 -s "from pathlib import PurePath;" "str(PurePath())"
500000 loops, best of 5: 1.97 usec per loop  # before
500000 loops, best of 5: 1.85 usec per loop  # after

$ ./python -m timeit -n 500000 -s "from pathlib import PurePath;" "str(PurePath('foo'))"
500000 loops, best of 5: 3.08 usec per loop  # before
500000 loops, best of 5: 2.96 usec per loop  # after

$ ./python -m timeit -n 500000 -s "from pathlib import PurePath;" "str(PurePath('foo', ''))"
500000 loops, best of 5: 4.08 usec per loop  # before
500000 loops, best of 5: 3.09 usec per loop  # after

$ ./python -m timeit -n 500000 -s "from pathlib import PurePath;" "str(PurePath('foo', 'bar'))"
500000 loops, best of 5: 4.45 usec per loop  # before
500000 loops, best of 5: 4.39 usec per loop  # after

$ ./python -m timeit -n 100000 -s "from pathlib import PurePath" "p = PurePath('a', 'b', 'c', 'd'); str(p); str(p / 'e')"
100000 loops, best of 5: 11.5 usec per loop  # before
100000 loops, best of 5: 10.5 usec per loop  # after
```

File: Lib/pathlib/_abc.py [hunk 1]
```diff
@@ -206,7 +206,11 @@ class PurePathBase:
     )
     pathmod = os.path
 
-    def __init__(self, *paths):
+    def __init__(self, *args):
+        paths = []
```
Comment (eendebakpt) [inline Lib/pathlib/_abc.py]: Can you use a list comprehension here? E.g `paths = [path for path in args if path]`. Avoiding the call to append might be faster.

Comment (barneygale) [reply Lib/pathlib/_abc.py]: Yes you're right.

I wrote it this way because I'm anticipating adding an `isinstance(arg, str)` check, but its premature.

**Comment by barneygale** [comment]
Timings after [28049ce](https://github.com/python/cpython/pull/113375/commits/28049ce1248af749cb3ee9c22a00a28ed92b8a22):

```shell
$ ./python -m timeit -n 500000 -s "from pathlib import PurePath;" "str(PurePath())"
500000 loops, best of 5: 1.97 usec per loop  # before
500000 loops, best of 5: 1.11 usec per loop  # after
# --> 1.77x faster

$ ./python -m timeit -n 500000 -s "from pathlib import PurePath;" "str(PurePath('foo'))"
500000 loops, best of 5: 3.08 usec per loop  # before
500000 loops, best of 5: 2.45 usec per loop  # after
# --> 1.26x faster

$ ./python -m timeit -n 500000 -s "from pathlib import PurePath;" "str(PurePath('foo', ''))"
500000 loops, best of 5: 4.08 usec per loop  # before
500000 loops, best of 5: 2.59 usec per loop  # after
# --> 1.58x faster

$ ./python -m timeit -n 500000 -s "from pathlib import PurePath;" "str(PurePath('foo', 'bar'))"
500000 loops, best of 5: 4.45 usec per loop  # before
500000 loops, best of 5: 3.74 usec per loop  # after
# --> 1.19x faster

$ ./python -m timeit -n 100000 -s "from pathlib import PurePath" "p = PurePath('a', 'b', 'c', 'd'); str(p); str(p / 'e')"
100000 loops, best of 5: 11.5 usec per loop  # before
100000 loops, best of 5: 9.28 usec per loop  # after
# --> 1.24x faster
```

File: Lib/pathlib/_abc.py [hunk 2]
```diff
@@ -206,8 +206,8 @@ class PurePathBase:
     )
     pathmod = os.path
 
-    def __init__(self, *paths):
-        self._raw_paths = paths
+    def __init__(self, *args):
+        self._raw_paths = [arg for arg in args if arg]
```
Comment (serhiy-storchaka) [inline Lib/pathlib/_abc.py]: It was a tuple, and now it is a list. Also, does not this processing add an overhead? In common case there should not be empty components (except maybe `Path('')`).

Comment (barneygale) [reply Lib/pathlib/_abc.py]: It add a tiny bit of overhead in the common case. Passing an empty string does happen, e.g. importlib calls `joinpath('')`, but perhaps not commonly enough to make this worthwhile. I'll undo.

File: Lib/pathlib/_abc.py [hunk 3]
```diff
@@ -205,8 +205,8 @@ class PurePathBase:
     )
     pathmod = posixpath
 
-    def __init__(self, *paths):
-        self._raw_paths = paths
+    def __init__(self, *args):
+        self._raw_paths = list(args)
```
Comment (serhiy-storchaka) [inline Lib/pathlib/_abc.py:209]: Why should it be a list?

Comment (barneygale) [reply Lib/pathlib/_abc.py:209]: `PurePath._raw_paths` is a list. `PurePathBase` has only just been split off from `PurePath`, and in doing so, I mistakenly changed the type of `_raw_paths` to a tuple, but only in `PurePathBase` and not `PurePath`. So the answer is: for consistency with its subclass, and to correct a mistake.

You could reasonably ask "why is `PurePath._raw_paths` a list?". The answer: `PurePath.__init__()` builds up the path list incrementally, and performs some special handling of other `PurePath` objects passed in as arguments. It's cheaper to use the list produced there as-is than convert it to a tuple/similar.

You could also ask "why fix it here?". The answer: this patch manipulates the list (see the "Store the joined path" bit of the patch), and so the difference between the list and tuple matters for the first time.

Comment (barneygale) [reply Lib/pathlib/_abc.py:209]: Perhaps I should add a comment here to explain why we convert it to a list?

**Comment by serhiy-storchaka** [comment]
I slightly concerned about increasing coupling and possible performance regressions in some scenarios, but the current design of pathlib is very different from the one that I know, so I have no confidence to either accept or reject these changes.

**Comment by barneygale** [comment]
Thanks for the review @serhiy-storchaka, very much appreciated. Given your concerns I'll withdraw this PR and have a re-think.

**Pull request closed without merging**